### PR TITLE
Multiple Updates

### PR DESCRIPTION
**Pull request by appukuttan-shailesh**
Multiple updates such as:
- updates authentication process
- adds helper methods to BaseClient
- adds a utils.generate_report() method using above
- updates requirements.txt in view of above
- expands documentation
- small fixes

### CODE REVIEW

File: hbp_validation_framework/__init__.py [hunk 1]
```diff
@@ -60,7 +50,15 @@ class BaseClient(object):
 
     def __init__(self, username,
                  password=None,
-                 url=VALIDATION_FRAMEWORK_URL):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py:63]: I don't like this change. We need to be able to support different URLs and different client ids, for example when running a local development server, beyond the two cases here.

I suggest adding client_id as a keyword argument to the constructor, with a default value of the production server, and then have an optional `config.json` file in the project directory where you can read alternative values for the url and client id (could also be used for other project-level configuration, like a regex or glob expression for which output files get uploaded to storage).

File: hbp_validation_framework/__init__.py [hunk 2]
```diff
@@ -180,6 +194,59 @@ def _hbp_auth(self, username, password):
         else:
             raise Exception("Something went wrong. Status code {} from NMPI, expected 302".format(rNMPI1.status_code))
 
+    def _translate_URL_to_UUID(self, path):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py:200]: This method should be in the `CollabDataStore` class.

File: hbp_validation_framework/__init__.py [hunk 3]
```diff
@@ -180,6 +194,59 @@ def _hbp_auth(self, username, password):
...
+        if path.startswith("collab://"):
+            path = path[len("collab://"):]
+        url = base_url + "?path=" + path
+        data = requests.get(url, auth=self.auth)
+        if data.status_code == 200:
+            return data.json()["uuid"]
+        else:
+            raise Exception("The provided 'path' is invalid! Error: " + str(data))
+
+    def _download_resource(self, uuid):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py:217]: This method should be in the `CollabDataStore` class.

File: hbp_validation_framework/__init__.py [hunk 4]
```diff
@@ -212,9 +279,10 @@ class TestLibrary(BaseClient):
     password : string, optional
         Your HBP Collaboratory password; advisable to not enter as plaintext.
         If left empty, you would be prompted for password at run time (safer).
-    url : string, optional
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py:215]: Same comment as above. We need to keep `url` field.

File: hbp_validation_framework/utils.py [hunk 5]
```diff
@@ -20,7 +21,8 @@
 from datetime import datetime
 from . import TestLibrary, ModelCatalog
 from .datastores import CollabDataStore
-
+from fpdf import FPDF
```
Comment (apdavison) [inline hbp_validation_framework/utils.py:24]: I suggest making these dependencies optional, i.e. `generate_report()` produces an error message saying "Please install these packages ...." if they are not installed.

**Review by apdavison — CHANGES_REQUESTED** [review]

**Review by apdavison — APPROVED** [review]

File: hbp_validation_framework/__init__.py (hunk 1 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py:63]: Resolved by introducing `environment` in d5ba8248ef8cd61e8f29f53b75d65e78ee6358a7 and planned feature in https://github.com/apdavison/hbp-validation-client/issues/15

File: hbp_validation_framework/__init__.py (hunk 2 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py:200]: Will be done in https://github.com/apdavison/hbp-validation-client/issues/17

File: hbp_validation_framework/__init__.py (hunk 3 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py:217]: Will be done in https://github.com/apdavison/hbp-validation-client/issues/17

File: hbp_validation_framework/__init__.py (hunk 4 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py:215]: Resolved by introducing `environment` in d5ba8248ef8cd61e8f29f53b75d65e78ee6358a7 and planned feature in https://github.com/apdavison/hbp-validation-client/issues/15

File: hbp_validation_framework/utils.py (hunk 5 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py:24]: Will be done in https://github.com/apdavison/hbp-validation-client/issues/18

**Pull request merged**